api/models/User.js

```diff
@@ -1,21 +1,97 @@
-const mongoose = require('mongoose');
-const Schema = mongoose.Schema;
-
- const userSchema = new Schema({
-     name: {
-         type: String,
-         required: true
-     },
-     email: {
-         type: String,
-         required: true
-     },
-     password: {
-         type: String,
-         required: true
-     }
- },{timestamps: true})
-
- const User = mongoose.model('User', userSchema)
-
- module.exports = User;
\ No newline at end of file
+var mongoose = require('mongoose');
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const confiq = require('../config/config').get(process.env.NODE_ENV);
+const salt = 10;
+
+const userSchema = mongoose.Schema({
+    name: {
+        type: String,
+        required: true,
+        unique: true,
+
+    },
+
+    email: {
+        type: String,
+        required: true,
+        trim: true,
+        unique: true
+    },
+    password: {
+        type: String,
+        required: true,
+        minlength: 6
+    },
+
+    token: {
+        type: String
+    }
+});
+// to signup a user
+userSchema.pre('save', function(next) {
+    var user = this;
+
+    if (user.isModified('password')) {
+        bcrypt.genSalt(salt, function(err, salt) {
+            if (err) return next(err);
+
+            bcrypt.hash(user.password, salt, function(err, hash) {
+                if (err) return next(err);
+                user.password = hash;
+
+                next();
+            })
+
+        })
+    } else {
+        next();
+    }
+});
+
+//to login
+userSchema.methods.comparepassword = function(password, cb) {
+    bcrypt.compare(password, this.password, function(err, isMatch) {
+        if (err) return cb(next);
+        cb(null, isMatch);
+    });
+}
+
+// generate token
+
+userSchema.methods.generateToken = function(cb) {
+    var user = this;
+    var token = jwt.sign(user._id.toHexString(), confiq.SECRET);
+
+    user.token = token;
+    user.save(function(err, user) {
+        if (err) return cb(err);
+        cb(null, user);
+    })
+}
+
+// find by token
+userSchema.statics.findByToken = function(token, cb) {
+    var user = this;
+
+    jwt.verify(token, confiq.SECRET, function(err, decode) {
+        user.findOne({ "_id": decode, "token": token }, function(err, user) {
+            if (err) return cb(err);
+            cb(null, user);
+        })
+    })
+};
+
+//delete token
+
+userSchema.methods.deleteToken = function(token, cb) {
+    var user = this;
+
+    user.update({ $unset: { token: 1 } }, function(err, user) {
+        if (err) return cb(err);
+        cb(null, user);
+    })
+}
+
+
```

Comment:
I am not sure if all of this belongs to here... Normally the model file is just a model for the dB, all the rest related with the model is done in external files.

Reply:
I would suggest the same, in a big application.
You don't want to mix auth logic and token within the model those are 2 different concepts.

However, this is a small enough application and I don't see the problem here. 
Mongoose provides this easy-to-use way for this sorta use case.